compiler/rustc_mir_transform/src/coverage/mod.rs

```diff
@@ -346,20 +346,15 @@ fn get_body_span<'tcx>(
     let mut body_span = hir_body.value.span;
 
     if tcx.is_closure(def_id.to_def_id()) {
-        // If the MIR function is a closure, and if the closure body span
-        // starts from a macro, but it's content is not in that macro, try
-        // to find a non-macro callsite, and instrument the spans there
-        // instead.
-        loop {
-            let expn_data = body_span.ctxt().outer_expn_data();
-            if expn_data.is_root() {
-                break;
-            }
-            if let ExpnKind::Macro { .. } = expn_data.kind {
-                body_span = expn_data.call_site;
-            } else {
-                break;
-            }
+        // If the current function is a closure, and its "body" span was created
+        // by macro expansion or async desugaring, try to walk backwards to the
+        // pre-expansion call site or body.
+        while let expn_data = body_span.ctxt().outer_expn_data()
+            && !expn_data.is_root()
```

Comment:
Redundant condition, `ExpnKind::Root` is never compatible with the second condition below.

Reply:
On the other hand, closure spans are never generated by non `Async` desugarings (?), so we can remove the second condition instead and always walk until encountering the root (maybe there's a readily available function for that in rustc_span).

Reply:
That method would be `Span::source_callsite`, but switching to it would also change the treatment of closure bodies produced by attribute macros and derive macros.

The existing test suite doesn't tell me whether that change would be better or worse, and it's beyond the scope of what I'm trying to achieve with this PR.

I'd prefer to stick with the targeted approach, only walking back through a narrower and more explicit subset of expansions (bang macros and async desugaring).

Reply:
(Deleted my previous comment; I was confusing this code with some similar code elsewhere.)

Reply:
I'm a little bit nervous of this change, but the cases I was worried about (non-bang macros) were already unexpanded by the pre-existing code, and the tests pass. I'm willing to give the suggested simplification a try.